### PR TITLE
3635 multi form alerts analytics

### PR DESCRIPTION
**Pull request by estellecomment**
# Description

When `multi_form_alert` is triggered, we generate alert messages on the `data_record`. Add fields on the corresponding `task` to make these alerts differentiable by analytics : `type:alert`, `alert_name`, and `countedReports`.

Note that `countedReports[0]` is the uuid of current report, and `countedReports[1..end]` are the uuids of the reports counted towards the alert threshold, sorted by submission date, most recent first.

medic/medic-webapp#3635

# Review checklist

- [x] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [x] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [x] Tested: Unit and/or e2e where appropriate
- [x] Internationalised: All user facing text
- [x] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.
- [ ] Webapp CI runs clean against this branch

### CODE REVIEW

**Comment by estellecomment** [comment]
The first 4 commits are cleanup/refactoring, the last 2 are the new code.

**Comment by estellecomment** [comment]
+ 1 bugfix on `lineage.hydrateReports`. 
I'll be squashing only partially for the merge, for clarity.

File: test/unit/default_responses.js [hunk 1]
```diff
@@ -370,28 +370,3 @@ exports['add response when recipient is allowed'] = function (test) {
         test.done();
     });
 };
-
-exports['add response with denied state when recipient is denied'] = function (test) {
```
Comment (garethbowen) [inline test/unit/default_responses.js]: This looks useful to me - why is this deleted?

File: test/unit/lineage.js [hunk 2]
```diff
@@ -466,22 +466,41 @@ exports['hydrate+minify is noop on a place'] = test => {
...
+exports['hydrateDocs works on docs without contacts or parents'] = test => {
+  const docs = [
+    { _id: 'a' },
+    { _id: 'b' },
+  ];
+  lineage.hydrateDocs(docs).then((hydratedDocs) => {
+    test.deepEqual(hydratedDocs, docs);
+    test.done();
+  });
+};
```
Comment (garethbowen) [inline test/unit/lineage.js]: Nice.

File: test/unit/transitions/multi_form_alerts.js [hunk 3]
```diff
@@ -163,18 +168,24 @@ exports['if no reports in time window, does nothing'] = test => {
   });
 };
 
-const assertMessage = (test, messageArgs, recipient, message) => {
-  test.deepEqual(messageArgs, {
+const assertMessage = (test, messageArgs, recipient, message, alertName) => {
+  let expected = {
```
Comment (garethbowen) [inline test/unit/transitions/multi_form_alerts.js]: `const` surely...

File: transitions/multi_form_alerts.js [hunk 4]
```diff
@@ -131,28 +137,28 @@ const validateConfig = () => {
     });
     alert.timeWindowInDays = parseInt(alert.timeWindowInDays);
     if (isNaN(alert.timeWindowInDays)) {
-      errors.push(`Alert number ${idx}, expecting "timeWindowInDays" to be an integer, eg: "timeWindowInDays": "3"`);
+      errors.push(`Alert "${alert.name}", expecting "timeWindowInDays" to be an integer, eg: "timeWindowInDays": "3"`);
```
Comment (garethbowen) [inline transitions/multi_form_alerts.js]: Much better!

File: transitions/multi_form_alerts.js [hunk 5]
```diff
@@ -131,28 +137,28 @@ const validateConfig = () => {
...
+      logger.warn(`Bad config for ${TRANSITION_NAME}, alert "${alert.name}". Expecting "forms" to be an array of form codes. Continuing without "forms", since it\'s optional.`);
     }
   });
   if (errors.length) {
-    logger.error(`Validation failed for ${NAME} transition`);
+    logger.error(`Validation failed for ${TRANSITION_NAME} transition`);
     logger.error(errors.join('\n'));
-    throw new Error(`Validation failed for ${NAME} transition`);
+    throw new Error(`Validation failed for ${TRANSITION_NAME} transition`);
   }
```
Comment (garethbowen) [inline transitions/multi_form_alerts.js]: Add validation to ensure all the `name` fields are unique.

File: test/unit/default_responses.js (hunk 1 above)
Comment (estellecomment) [inline test/unit/default_responses.js]: because the code in `default_responses.js` actually doesn't do that behavior o_O. 
It's the `messages.js` that does the `denied` status, and it's not specific to default responses anyway. The test should have been broken, it was testing something impossible, but it was passing because it was using a spy instead of a mock.

**Comment by estellecomment** [comment]
yup, webapp PR is coming (forgot to push!)
